x-pack/plugins/inference/server/chat_complete/adapters/gemini/gemini_adapter.ts

```diff
@@ -0,0 +1,234 @@
+                  : {
+                      type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+                      properties: {},
+                    },
+              };
+            }
+          ),
+        },
+      ]
+    : [];
+}
+
+function toolSchemaToGemini({ schema }: { schema: ToolSchema }): Gemini.FunctionDeclarationSchema {
+  const convertSchemaType = ({
+    def,
+  }: {
+    def: ToolSchemaType;
+  }): Gemini.FunctionDeclarationSchemaProperty => {
+    switch (def.type) {
+      case 'array':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.ARRAY,
+          description: def.description,
+          items: convertSchemaType({ def: def.items }) as Gemini.FunctionDeclarationSchema,
+        };
+      case 'object':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+          description: def.description,
+          required: def.required as string[],
+          properties: Object.entries(def.properties).reduce<
+            Record<string, Gemini.FunctionDeclarationSchema>
+          >((properties, [key, prop]) => {
+            properties[key] = convertSchemaType({ def: prop }) as Gemini.FunctionDeclarationSchema;
+            return properties;
+          }, {}),
+        };
+      case 'string':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.STRING,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+      case 'boolean':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.BOOLEAN,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+      case 'number':
+        return {
+          type: Gemini.FunctionDeclarationSchemaType.NUMBER,
+          description: def.description,
+          enum: def.enum ? (def.enum as string[]) : def.const ? [def.const] : undefined,
+        };
+    }
+  };
+
+  return {
+    type: Gemini.FunctionDeclarationSchemaType.OBJECT,
+    required: schema.required as string[],
+    properties: Object.entries(schema.properties).reduce<
+      Record<string, Gemini.FunctionDeclarationSchemaProperty>
+    >((properties, [key, def]) => {
+      properties[key] = convertSchemaType({ def });
+      return properties;
+    }, {}),
+  };
+}
+
+function messagesToGemini({
+  messages,
+  system,
+}: {
+  messages: Message[];
+  system?: string;
+}): GeminiMessage[] {
+  // systemInstruction is not supported on all gemini versions
+  // so for now we just always use the old trick of user message + assistant acknowledge.
+  // See https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/inference#request
+  const systemMessages: GeminiMessage[] | undefined = system
+    ? [
+        { role: 'user', content: system },
+        { role: 'assistant', content: 'Understood.' },
+      ]
+    : undefined;
```

Comment:
`systemInstruction` is available for [gemini-1.5-flash, gemini-1.5-pro, and gemini-1.0-pro-002](https://cloud.google.com/vertex-ai/generative-ai/docs/model-reference/inference#request). This is slightly less support than for function invocation (as gemini.1.0-pro-001 isn't in the list), so I kept the hacky user-system-message approach for now, but we might use `systemInstruction` very soon, as the version not supporting it are discontinued in less than 6 months.

Reply:
I would not be against dropping this entirely. I think we've mostly been talking about gemini 1.5 (e.g. in https://www.elastic.co/blog/whats-new-elastic-observability-8-15-0 and https://www.elastic.co/blog/whats-new-elastic-security-8-15-0).